moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java

```diff
@@ -0,0 +1,193 @@
+      var replaced = bindings.put(jsonName, componentBinding);
+      if (replaced != null) {
+        throw new IllegalArgumentException(
+            "Conflicting components:\n"
+                + "    "
+                + replaced.name
+                + "\n"
+                + "    "
+                + componentBinding.name);
+      }
+    }
+    Constructor<Object> constructor;
+    try {
+      //noinspection unchecked
+      constructor = (Constructor<Object>) rawType.getDeclaredConstructor(constructorParams);
+      constructor.setAccessible(true);
+    } catch (NoSuchMethodException e) {
+      throw new AssertionError(e);
+    }
+    return new RecordJsonAdapter<>(constructor, bindings).nullSafe();
+  };
+
+  private static record ComponentBinding<T>(String name, String jsonName, JsonAdapter<T> adapter, Method accessor) {
+  }
+
+    private final String targetClass;
+    private final Constructor<T> constructor;
+    private final ComponentBinding<Object>[] componentBindingsArray;
+    private final JsonReader.Options options;
+
+    @SuppressWarnings("ToArrayCallWithZeroLengthArrayArgument")
+    public RecordJsonAdapter(
+        Constructor<T> constructor,
+        Map<String, ComponentBinding<?>> componentBindings
+    ) {
+      this.constructor = constructor;
+      this.targetClass = constructor.getDeclaringClass().getSimpleName();
+      //noinspection unchecked
+      this.componentBindingsArray = componentBindings.values().toArray(new ComponentBinding[componentBindings.size()]);
+      this.options = JsonReader.Options.of(componentBindings.keySet().toArray(new String[componentBindings.size()]));
+    }
+
+    @Override
+    public T fromJson(JsonReader reader) throws IOException {
+      var resultsArray = new Object[componentBindingsArray.length];
+
+      reader.beginObject();
+      while (reader.hasNext()) {
+        int index = reader.selectName(options);
+        if (index == -1) {
+          reader.skipName();
+          reader.skipValue();
+          continue;
+        }
+        var result = componentBindingsArray[index].adapter.fromJson(reader);
+        resultsArray[index] = result;
+      }
+      reader.endObject();
+
+      var initialAccess = constructor.canAccess(null);
+      try {
+        if (!initialAccess) {
+          constructor.setAccessible(true);
+        }
+        return constructor.newInstance(resultsArray);
+      } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
+        throw new AssertionError(e);
+      } finally {
+        constructor.setAccessible(false);
+      }
+    }
+
+    @Override
+    public void toJson(JsonWriter writer, T value) throws IOException {
+      writer.beginObject();
+
+      for (var binding : componentBindingsArray) {
+        writer.name(binding.jsonName);
+        try {
+          binding.adapter.toJson(writer, binding.accessor.invoke(value));
+        } catch (IllegalAccessException | InvocationTargetException e) {
```

Comment:
`InvocationTargetException` should be unwrapped and re-thrown since it indicates an exception on the other side of the reflection.
```java
catch (InvocationTargetException e) {
  Throwable cause = e.getCause();
  if (cause instanceof RuntimeException) throw (RuntimeException) cause;
  if (cause instanceof Error) throw (Error) cause;
  throw new RuntimeException(cause);
}
```

Reply:
765a642 and also part of later changes